common/network-common/src/main/java/org/apache/spark/network/sasl/aes/SparkAesCipher.java

```diff
@@ -0,0 +1,270 @@
+import java.security.NoSuchAlgorithmException;
+import java.util.Arrays;
+import java.util.Properties;
+import javax.crypto.Cipher;
+import javax.crypto.Mac;
+import javax.crypto.SecretKey;
+import javax.crypto.ShortBufferException;
+import javax.crypto.spec.SecretKeySpec;
+import javax.crypto.spec.IvParameterSpec;
+import javax.security.sasl.SaslException;
+
+import org.apache.commons.crypto.cipher.CryptoCipher;
+import org.apache.commons.crypto.utils.Utils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * AES cipher for encryption and decryption.
+ */
+public class SparkAesCipher {
+  private static final Logger logger = LoggerFactory.getLogger(SparkAesCipher.class);
+  private static final byte[] EMPTY_BYTE_ARRAY = new byte[0];
+  public final static String supportedTransformation[] = {
+    "AES/CBC/NoPadding", "AES/CTR/NoPadding"
+  };
+
+  private final CryptoCipher encryptor;
+  private final CryptoCipher decryptor;
+
+  private final Integrity integrity;
+
+  public SparkAesCipher(
+      String cipherTransformation,
+      Properties properties,
+      byte[] inKey,
+      byte[] outKey,
+      byte[] inIv,
+      byte[] outIv) throws IOException {
+    if (!Arrays.asList(supportedTransformation).contains(cipherTransformation)) {
+      logger.warn("AES cipher transformation is not supported: " + cipherTransformation);
+      cipherTransformation = "AES/CTR/NoPadding";
+      logger.warn("Use default AES/CTR/NoPadding");
+    }
+
+    final SecretKeySpec inKeySpec = new SecretKeySpec(inKey, "AES");
+    final IvParameterSpec inIvSpec = new IvParameterSpec(inIv);
+    final SecretKeySpec outKeySpec = new SecretKeySpec(outKey, "AES");
+    final IvParameterSpec outIvSpec = new IvParameterSpec(outIv);
+
+    // Encryptor
+    encryptor = Utils.getCipherInstance(cipherTransformation, properties);
+    try {
+      logger.debug("Initialize encryptor");
+      encryptor.init(Cipher.ENCRYPT_MODE, outKeySpec, outIvSpec);
+    } catch (InvalidKeyException | InvalidAlgorithmParameterException e) {
+      throw new IOException("Failed to initialize encryptor", e);
+    }
+
+    // Decryptor
+    decryptor = Utils.getCipherInstance(cipherTransformation, properties);
+    try {
+      logger.debug("Initialize decryptor");
+      decryptor.init(Cipher.DECRYPT_MODE, inKeySpec, inIvSpec);
+    } catch (InvalidKeyException | InvalidAlgorithmParameterException e) {
+      throw new IOException("Failed to initialize decryptor", e);
+    }
+
+    integrity = new Integrity(outKey, inKey);
+  }
+
+  /**
+   * Encrypts input data. The result composes of (msg, padding if needed, mac) and sequence num.
+   * @param data the input byte array
+   * @param offset the offset in input where the input starts
+   * @param len the input length
+   * @return the new encrypted byte array.
+   * @throws SaslException if error happens
+   */
+  public byte[] wrap(byte[] data, int offset, int len) throws SaslException {
```

Comment:
I suppose you were talking about SaslEncrytionbackend. Actually, the real underlying interfaces of wrap/unwrap are defined by SaslClient which provided by java security.